test_conformance/profiling/profiling_timebase.cpp

```diff
@@ -0,0 +1,92 @@
+//
+// Copyright (c) 2021 The Khronos Group Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include "procs.h"
+
+const char *kernelCode = "__kernel void kernel_empty(){}";
+
+int test_profiling_timebase(cl_device_id device, cl_context context,
+                            cl_command_queue queue, int num_elements)
+{
+    cl_int err;
+    cl_ulong hostTime;
+    clProgramWrapper program;
+    clKernelWrapper kernel;
+    clEventWrapper kEvent;
+
+    clEventWrapper uEvent = clCreateUserEvent(context, &err);
+    test_error(err, "Failed to create user event");
+
+    err = create_single_kernel_helper(context, &program, &kernel, 1,
+                                      &kernelCode, "kernel_empty");
+    test_error(err, "Failed to create kernel");
+
+    cl_ulong deviceTimeBeforeQueue;
+    err = clGetDeviceAndHostTimer(device, &deviceTimeBeforeQueue, &hostTime);
+    test_error(err, "Unable to get starting device time");
+
+    clEventWrapper userEvents = uEvent;
+    err = clEnqueueNDRangeKernel(queue, kernel, 1, NULL, NULL, NULL, 1,
+                                 &userEvents, &kEvent);
+    test_error(err, "clEnqueueNDRangeKernel failed");
+
+    cl_ulong deviceTimeAfterQueue;
+    err = clGetDeviceAndHostTimer(device, &deviceTimeAfterQueue, &hostTime);
+    test_error(err, "Unable to get queue device time");
+
+    err = clFlush(queue);
+    test_error(err, "clFlush failed");
+
+    err = clSetUserEventStatus(uEvent, CL_COMPLETE);
+    test_error(err, "Unable to complete user event");
+
+    err = clWaitForEvents(1, &kEvent);
+    test_error(err, "clWaitForEvents failed");
+
+    cl_ulong deviceTimeAfterCompletion;
+    err =
+        clGetDeviceAndHostTimer(device, &deviceTimeAfterCompletion, &hostTime);
```

Comment:
This is optional feature in OpenCL 3.0. You cannot assume this function is always available
https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_device_and_host_timer_synchronization